allure-testng-adaptor/src/main/java/ru/yandex/qatools/allure/testng/AllureTestListener.java

```diff
@@ -41,48 +37,103 @@
  * @author Dmitry Baev charlie@yandex-team.ru
  *         Date: 22.11.13
  */
-public class AllureTestListener implements ITestListener, IConfigurationListener {
+public class AllureTestListener implements IResultListener, ISuiteListener {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(AllureTestListener.class);
 
     private Allure lifecycle = Allure.LIFECYCLE;
 
-    private Map<String, String> suiteUid = new HashMap<>();
-
-    private Set<String> startedTestNames = Collections.newSetFromMap(
-            new ConcurrentHashMap<String, Boolean>());
+    private static final String SUITE_UID = "SUITE_UID"; 
+    
+    @Override
+    public void onStart(ISuite suite) {
+        if (suite.getAttribute(SUITE_UID) == null) {
+            String uid = UUID.randomUUID().toString();
+            suite.setAttribute(SUITE_UID, uid);
+        }
+    }
 
     @Override
+    public void onFinish(ISuite suite) {
+        if (suite.getAttribute(SUITE_UID) != null) {
+            String uid = (String) suite.getAttribute(SUITE_UID);
+            suite.removeAttribute(SUITE_UID);
+            getLifecycle().fire(new TestSuiteFinishedEvent(uid));
+        }
+    }
+    
+    @Override
     public void onConfigurationSuccess(ITestResult iTestResult) {
         //nothing
     }
 
     @Override
     public void onConfigurationFailure(ITestResult iTestResult) {
-        if (!startedTestNames.contains(getName(iTestResult))) {
-            onTestStart(iTestResult);
-        }
-
         Throwable throwable = iTestResult.getThrowable();
-        if (throwable == null) {
-            throwable = new TestException("Test configuration failure");
-        }
-
-        getLifecycle().fire(new TestCaseCanceledEvent()
-                        .withThrowable(throwable)
-        );
+        createConfigEvent(iTestResult);
+        getLifecycle().fire(new TestCaseFailureEvent().withThrowable(throwable));
         fireFinishTest();
     }
 
     @Override
     public void onConfigurationSkip(ITestResult iTestResult) {
-        //nothing
+        createConfigEvent(iTestResult);
+        fireTestCaseCancel(iTestResult);
+        fireFinishTest();
     }
 
+    private void createConfigEvent(ITestResult iTestResult) {
+        String description = iTestResult.getMethod().getDescription();
+        if (description==null || description.trim().isEmpty()) {
```

Comment:
trim can be removed if you wish. It's just to validate that condition is true when the description have white spaces like " "